### PR TITLE
CI: Add GHA workflow to upload nightly wheels

### LINKED ISSUE #9994: document where nightly wheels are published
It would be great to have a nightly pip package as provided by [tensorflow](https://pypi.python.org/pypi/tf-nightly) such that users can easily experiment with the most recent code without having to build from source, e.g. publish a package `matplotlib-nightly` to `pypi` as part of builds on `master`. 

I came across [this](https://github.com/matplotlib/matplotlib/blob/38be7aeaaac3691560aeadafe46722dda427ef47/doc/devel/MEP/MEP19.rst#requirements) document from 2015 discussing a similar idea, but couldn't find anything on a nightly package in the [installation instructions](http://matplotlib.org/devdocs/users/installing.html).

### LINKED ISSUE #21635: [ENH]: Add a nightly wheel build
### Problem

Automated testing of third party packages against the latest matplotlib currently requires installing from source which takes quite a while.

### Proposed solution

Add a github action (with cibuildwheel maybe?) that builds the latest matplotlib on one or a few OSes x python versions. Somethign like:

Python 3.8 and 3.9 on latest ubuntu

These wheels should then be uploaded to a consistent place (maybe somewhere on maptlotlib.org?) with a clear naming scheme so it's easy to install then from a scheduled github action.

### PR DESCRIPTION
**Pull request by matthewfeickert**
## PR Summary

* Resolves #21635
* Resolves #9994
* Closes #21637 ([supersedes](https://github.com/matplotlib/matplotlib/pull/21637#issuecomment-1082582723))

This PR adds a `nightlies.yml` GitHub Action workflow that uses the `gh` CLI API to download the 'wheels' artifact from the latest completed build of the `cibuildwheel` workflow from the 'main' branch (idea taken from https://github.com/actions/download-artifact/issues/3#issuecomment-1017141067).

<details>
<summary>gh run example:</summary>

```console
$ gh run --repo matplotlib/matplotlib list --branch main --workflow cibuildwheel.yml
STATUS  NAME                                                                  WORKFLOW         BRANCH  EVENT         ID          ELAPSED   AGE
✓       Merge pull request #22719 from oscargus/incorrectdeprecationwarning   Build CI wheels  main    push          2067808346  1h11m7s   5h
✓       Merge pull request #22138 from stanleyjs/subfigure-clear              Build CI wheels  main    push          2067068506  1h11m8s   7h
✓       Merge pull request #22698 from oscargus/fixdoclinks                   Build CI wheels  main    push          2063032633  1h24m7s   21h
✓       Merge pull request #22711 from andrew-fennell/RangeSlider-bugfix      Build CI wheels  main    push          2062983684  1h25m25s  21h
✓       Merge pull request #22263 from jklymak/doc-version-switcher-condense  Build CI wheels  main    push          2061330182  1h14m20s  1d
✓       Merge pull request #22361 from anntzer/datetex                        Build CI wheels  main    push          2061179011  1h18m39s  1d
✓       Merge pull request #22721 from anntzer/style                          Build CI wheels  main    push          2059315555  1h3m24s   1d
X       PendingDeprecationWarning:                                            Build CI wheels  main    pull_request  2059201532  0s        1d
X       change test_aspect_equal_error() into test_aspect_equal()             Build CI wheels  main    pull_request  2057642015  0s        1d
✓       Merge pull request #22356 from timhoffm/triplot                       Build CI wheels  main    push          2057490803  1h2m25s   1d
✓       Merge pull request #22360 from anntzer/multilinetex                   Build CI wheels  main    push          2057057079  1h3m4s    1d
✓       Merge pull request #22418 from anntzer/ov                             Build CI wheels  main    push          2057027071  1h0m11s   1d
✓       Merge pull request #22722 from anntzer/cmf                            Build CI wheels  main    push          2056818352  1h8m22s   1d
✓       Merge pull request #22697 from oscargus/removecleanuptestcase         Build CI wheels  main    push          2056286785  1h3m30s   1d
✓       Merge pull request #22716 from jklymak/doc-set-canonical              Build CI wheels  main    push          2056271472  1h1m49s   1d
✓       Merge pull request #22556 from oscargus/parse_math_rcparams           Build CI wheels  main    push          2056265621  1h15m47s  2d
-       Locale font                                                           Build CI wheels  main    pull_request  2051994021  4s        2d
-       Locale font                                                           Build CI wheels  main    pull_request  2051993409  3s        2d
✓       Merge pull request #22163 from daniilS/tk_toolbar_colours             Build CI wheels  main    push          2044974231  1h13m49s  4d
-       Provide axis('equal') for Axes3D                                      Build CI wheels  main    pull_request  2040650565  3s        5d

For details on a run, try: gh run view <run-id>
```

</details>

This is done by getting the fields `event`, `status`, and `databaseId` from the output of `gh run list` as JSON and then using `jq` to filter for 'push' events (which correspond to merged PRs to the 'main' branch) that have completed runs. This is then sorted by the `databaseId`s in descending order to get the latest completed build run number.

<details>
<summary>filter and download example:</summary>

```console
$ gh run --repo matplotlib/matplotlib list --branch main --workflow cibuildwheel.yml --json event,status,databaseId
[
  {
    "databaseId": 2067808346,
    "event": "push",
    "status": "completed"
  },
  {
    "databaseId": 2067068506,
    "event": "push",
    "status": "completed"
  },
  ...,
  {
    "databaseId": 2044974231,
    "event": "push",
    "status": "completed"
  },
  {
    "databaseId": 2040650565,
    "event": "pull_request",
    "status": "completed"
  }
]
$ cat runs.json | jq -c '[ .[] | select(.event == "push") | select(.status == "completed") ] | sort_by(.databaseId) | reverse'
[{"databaseId":2067808346,"event":"push","status":"completed"},{"databaseId":2067068506,"event":"push","status":"completed"},{"databaseId":2063032633,"event":"push","status":"completed"},{"databaseId":2062983684,"event":"push","status":"completed"},{"databaseId":2061330182,"event":"push","status":"completed"},{"databaseId":2061179011,"event":"push","status":"completed"},{"databaseId":2059315555,"event":"push","status":"completed"},{"databaseId":2057490803,"event":"push","status":"completed"},{"databaseId":2057057079,"event":"push","status":"completed"},{"databaseId":2057027071,"event":"push","status":"completed"},{"databaseId":2056818352,"event":"push","status":"completed"},{"databaseId":2056286785,"event":"push","status":"completed"},{"databaseId":2056271472,"event":"push","status":"completed"},{"databaseId":2056265621,"event":"push","status":"completed"},{"databaseId":2044974231,"event":"push","status":"completed"}]
$ cat runs.json | jq -c '[ .[] | select(.event == "push") | select(.status == "completed") ] | sort_by(.databaseId) | reverse | .[0].databaseId'
2067808346
$ gh run --repo matplotlib/matplotlib download 2067808346 --name wheels
$ ls *.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-macosx_10_12_universal2.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-macosx_10_12_x86_64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-macosx_11_0_arm64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-manylinux_2_17_i686.manylinux2014_i686.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-win32.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp310-cp310-win_amd64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-macosx_10_12_universal2.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-macosx_10_12_x86_64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-macosx_11_0_arm64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-manylinux_2_5_i686.manylinux1_i686.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-win32.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp38-cp38-win_amd64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-macosx_10_12_universal2.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-macosx_10_12_x86_64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-macosx_11_0_arm64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-manylinux_2_5_i686.manylinux1_i686.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-win32.whl
matplotlib-3.6.0.dev1935+gda9533d507-cp39-cp39-win_amd64.whl
```

</details>

The ~~subset of all the~~ downloaded wheels ~~that are `x86_64.manylinux*`~~ are then uploaded to the `scipy-wheels-nightly` organization using the `anaconda-client` CLI API.

N.B.: `anaconda-client` must currently be installed from GitHub as there have been no uploads to PyPI since 2016. c.f. https://github.com/Anaconda-Platform/anaconda-client/issues/540

The workflow runs nightly as a CRON job ('schedule' in GHA parlance) at 01:23 UTC (the time has no meaning (it is just 0123) beyond making sure it misses the wall of CRON jobs that slams CI providers at 00:00 and 00:01) and on demand through workflow dispatch.

The status of the build on my fork is currently summarized in https://github.com/matplotlib/matplotlib/issues/21635#issuecomment-1082777172.

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

(No Python code was changed, so checking these off as `main` is passing)

**Documentation**
- [x] New features are documented, with examples if plot related. (N/A as no new features)
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there). (N/A as no new features)
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there). (N/A as no API changes)
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [inline .github/workflows/nightlies.yml:6]: This can get changed to be anything that people would like. Chosen just to miss the `00:00` and `00:01` nightly rush.

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [inline .github/workflows/nightlies.yml:19]: Using `3.x` instead of fixing to `3.10` as this should work with any stable CPython release. Happy to change if needed.

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [inline .github/workflows/nightlies.yml:55]: c.f. https://github.com/matplotlib/matplotlib/pull/21637#issuecomment-971947384

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:57]: Okay great. I'll upload all of them and given that this is a simple thing and we're not deep into review I'm going to rebase this.

**Comment by matthewfeickert** [comment]
Okay, this passed CI before I rebased it to allow for uploading all the wheels (c.f. https://github.com/matplotlib/matplotlib/pull/22733#discussion_r838826484) so this should be ready for review. I'm going to tag multiple people as there as lots of discussion in Issue #9994 (apologies if this annoys anyone &mdash; not the intention).

Tagging:
* @matplotlib/developers as was suggested by the very nice welcome bot. :heart_eyes:
* @ianhi as they already have been thinking about this work a lot and lead it with PR #21637.
* @matthew-brett, @jklymak, @QuLogic given discussion and input in Issue #9994 and PR #21637.
* @greglucas as they pointed out documentation is needed in https://github.com/matplotlib/matplotlib/pull/21637#issuecomment-972423545.
* @tacaswell as he'll need to verify that `ANACONDA_ORG_UPLOAD_TOKEN` is still correct in the repo secrets (c.f. https://github.com/matplotlib/matplotlib/pull/21637#issuecomment-971947384). Maybe he could do a manual run first to verify this?

File: .github/workflows/nightlies.yml
Comment (tacaswell) [reply .github/workflows/nightlies.yml:50]: `ANACONDA_ORG_UPLOAD_TOKEN` is still the right secret but I do not think we will know if this works until we merge it.

**Review by tacaswell — APPROVED** [review]

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:50]: Not sure if it is worth it, but you could manually download a wheel and then upload it with the token to test in advance that the `ANACONDA_ORG_UPLOAD_TOKEN` is valid. Though given that this PR enables workflow dispatch if there was something wrong with the token you could just change the value of the repo secret and rerun the job (which runs in under 1 minute) so probably not worth the time to test it out before.

File: .github/workflows/nightlies.yml
Comment (QuLogic) [inline .github/workflows/nightlies.yml:32]: Is this guaranteed to a) be sorted the way you assume, and b) not include PRs that target the branch? It is not obvious from the `gh run list` docs.

File: doc/users/installing/index.rst
Comment (QuLogic) [inline doc/users/installing/index.rst:90]:
```suggestion
These wheels can be installed with ``pip`` by specifying scipy-wheels-nightly
```

File: .github/workflows/nightlies.yml
Comment (QuLogic) [inline .github/workflows/nightlies.yml:13]: We probably don't want to run this on forks, so should add a condition here:
```suggestion
    runs-on: ubuntu-latest
    if: github.repository_owner == 'matplotlib'
```

File: .github/workflows/nightlies.yml
Comment (QuLogic) [inline .github/workflows/nightlies.yml:6]:
```suggestion
  schedule:
    - cron: '23 1 * * *'
```

File: .github/workflows/nightlies.yml
Comment (QuLogic) [inline .github/workflows/nightlies.yml:35]: I'd also wrap these at 80, though I'm not sure how best to wrap the `jq` script.
```suggestion
          gh run --repo "${PROJECT_REPO}" \
             list --branch "${BRANCH}" \
                  --workflow "${WORKFLOW_NAME}" \
                  --json status,databaseId > runs.json
          # Get latest completed build of wheels
          RUN_ID=$(cat runs.json | \
            jq -c '[ .[] | select(.status == "completed") ] | sort_by(.databaseId) | reverse | .[0].databaseId'
          )
          gh run --repo "${PROJECT_REPO}" \
             download "${RUN_ID}" --name "${ARTIFACT_NAME}"

```

File: .github/workflows/nightlies.yml
Comment (tacaswell) [reply .github/workflows/nightlies.yml:13]: Forks won't have access to the secret at least.

Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:13]: Yeah, this makes sense. We do this on certain workflows for pyhf with `github.repository == 'scikit-hep/pyhf'`.

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:32]: > Is this guaranteed to a) be sorted the way you assume

While I believe yes, in the RUN_ID line I explicitly do a sort in descending order of `databaseId` which is monotonically increasing in time so we don't have to hope.

> not include PRs that target the branch?

Hm. Good point! As I show in the "gh run example" in the PR body this can also pick up `pull_request` events. So I should rewrite this to explicitly filter on `event === "push"`.

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [reply .github/workflows/nightlies.yml:35]: > I'd also wrap these at 80

Okay. I'm going to have to rewrite this to deal with the `event` filter so I'll stick to 80 there. 

> though I'm not sure how best to wrap the `jq` script.

Here I personally think sticking to 80 makes the legibility significantly worse. If this is a hard rule then it would be better to split the `jq` steps up into multiple stages.

**Comment by matthewfeickert** [comment]
@QuLogic Question before I push next: Are you cool with rebasing during code review? Or would you prefer that the commit history not get revised until after final approval and then a final rebase be done to keep things more atomic and clean? The [dev docs seem neutral on this point](https://matplotlib.org/stable/devel/gitwash/development_workflow.html) but I want to make your review as easy and not annoying as possible. :+1:

**Comment by tacaswell** [comment]
We are very pro re-basing.

File: .github/workflows/nightlies.yml
Comment (matthewfeickert) [inline .github/workflows/nightlies.yml:39]: This line is now at 87. So still not at 80, but I think this is pretty close while still keeping readability.

File: .github/workflows/nightlies.yml
Comment (QuLogic) [inline .github/workflows/nightlies.yml:44]: No need for the subshell, and now that I just checked, no need for `cat` either (alas I don't see a way to combine the two `select` into one or we might have made it under 80 even, but it's close enough.)
```suggestion
          jq --compact-output \
            '[ .[] | select(.event == "push") | select(.status == "completed") ]' \
            runs.json > pushes.json
          # Get id of latest build of wheels
          RUN_ID=$(
            jq --compact-output \
              'sort_by(.databaseId) | reverse | .[0].databaseId' pushes.json
          )
```
Also using the full argument name will be clearer for people.

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @matthewfeickert! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Comment by QuLogic** [comment]
@tacaswell: The [token is incorrect](https://github.com/matplotlib/matplotlib/runs/5766379691?check_suite_focus=true):
```
Error:  ('Authentication token does not have the sufficient scope to perform this action expected: api:write', 401)
```

**Comment by matthewfeickert** [comment]
> Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

Thanks for a :zap: fast and helpful review! :smile: Truly appreciate all the help here &mdash; makes for a good first time contributor experience. :+1:

---

I guess the "good news" RE: the token scope problem is that this shouldn't require any code changes at least.

**Comment by tacaswell** [comment]
https://github.com/matplotlib/matplotlib/runs/5780119640?check_suite_focus=true <- like it works now.

**Comment by ianhi** [comment]
Looks like in CI the given command doesn't necessarily work. because it tries to pull in the matlpotlib dependencies from the same source. See https://github.com/matplotlib/ipympl/runs/5780345225?check_suite_focus=true

```
Looking in indexes: https://pypi.anaconda.org/scipy-wheels-nightly/simple
Collecting matplotlib
  Downloading https://pypi.anaconda.org/scipy-wheels-nightly/simple/matplotlib/3.6.0.dev1948%2Bgd8ede1a710/matplotlib-3.6.0.dev1948%2Bgd8ede1a710-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 59.0 MB/s eta 0:00:00
ERROR: Could not find a version that satisfies the requirement cycler>=0.10 (from matplotlib) (from versions: none)
ERROR: No matching distribution found for cycler>=0.10
Error: Process completed with exit code 1.
```

**Comment by QuLogic** [comment]
Perhaps it should be `--extra-index-url`?

**Comment by ianhi** [comment]
i think `extra` needs to point to pypi?

**Comment by ianhi** [comment]
follow along at https://github.com/matplotlib/ipympl/pull/449 for experimentation

**Comment by matthewfeickert** [comment]
> https://github.com/matplotlib/matplotlib/runs/5780119640?check_suite_focus=true <- like it works now.

YAY! :D 

matplotlib_wheels

@ianhi @QuLogic This is what I was assuming people would be doing (given [how we test `scipy` for `pyhf`](https://github.com/scikit-hep/pyhf/blob/0a8f3282e9bc6d66b3e596fd35c74a6e03ff567a/.github/workflows/dependencies-head.yml#L52-L60))

```console
$ docker run --rm -ti python:3.10 /bin/bash
root@1d34811c6d56:/# python -m venv venv && . venv/bin/activate
(venv) root@1d34811c6d56:/# python -m pip --quiet install --upgrade pip setuptools wheel
(venv) root@1d34811c6d56:/# python -m pip --quiet install matplotlib  # Get dependencies from PyPI
(venv) root@d092a0562fab:/# python -m pip install --upgrade --index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple matplotlib
Looking in indexes: https://pypi.anaconda.org/scipy-wheels-nightly/simple
Requirement already satisfied: matplotlib in /venv/lib/python3.10/site-packages (3.5.1)
Collecting matplotlib
  Downloading https://pypi.anaconda.org/scipy-wheels-nightly/simple/matplotlib/3.6.0.dev1948%2Bgd8ede1a710/matplotlib-3.6.0.dev1948%2Bgd8ede1a710-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 5.8 MB/s eta 0:00:00
Requirement already satisfied: packaging>=20.0 in /venv/lib/python3.10/site-packages (from matplotlib) (21.3)
Requirement already satisfied: numpy>=1.19 in /venv/lib/python3.10/site-packages (from matplotlib) (1.22.3)
Requirement already satisfied: fonttools>=4.22.0 in /venv/lib/python3.10/site-packages (from matplotlib) (4.31.2)
Requirement already satisfied: cycler>=0.10 in /venv/lib/python3.10/site-packages (from matplotlib) (0.11.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /venv/lib/python3.10/site-packages (from matplotlib) (1.4.2)
Requirement already satisfied: pillow>=6.2.0 in /venv/lib/python3.10/site-packages (from matplotlib) (9.0.1)
Requirement already satisfied: python-dateutil>=2.7 in /venv/lib/python3.10/site-packages (from matplotlib) (2.8.2)
Requirement already satisfied: pyparsing>=2.2.1 in /venv/lib/python3.10/site-packages (from matplotlib) (3.0.7)
Requirement already satisfied: six>=1.5 in /venv/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)

```

Though it seems that this isn't how you think people will approach it, so I guess this should get updated.

**Comment by tacaswell** [comment]
I think the pyparsing issue is one we should look into un-pinning on main.

**Comment by ianhi** [comment]
Looks like the magic working line is this one:

```
pip install --upgrade --index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple --extra-index-url https://pypi.org/simple matplotlib
```

**Comment by ianhi** [comment]
actually no it doesn't :/

still got `3.5.1`

https://github.com/matplotlib/ipympl/runs/5780465750?check_suite_focus=true

```
Looking in indexes: https://pypi.anaconda.org/scipy-wheels-nightly/simple, https://pypi.org/simple
Collecting matplotlib
  Downloading matplotlib-3.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 40.6 MB/s eta 0:00:00
```

**Comment by matthewfeickert** [comment]
> I think the pyparsing issue is one we should look into un-pinning on main.

Yeah this is necessary. You currently have to do the following:

```console
$ docker run --rm -ti python:3.10 /bin/bash
root@ad729996566a:/# python -m venv venv && . venv/bin/activate
(venv) root@ad729996566a:/# python -m pip --quiet install --upgrade pip setuptools wheel
(venv) root@ad729996566a:/# python -m pip --quiet install matplotlib 
(venv) root@ad729996566a:/# python -m pip --quiet uninstall -y matplotlib
(venv) root@ad729996566a:/# python -m pip --quiet install --upgrade 'pyparsing<3.0'
(venv) root@ad729996566a:/# python -m pip install --upgrade --index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple matplotlib
Looking in indexes: https://pypi.anaconda.org/scipy-wheels-nightly/simple
Collecting matplotlib
  Downloading https://pypi.anaconda.org/scipy-wheels-nightly/simple/matplotlib/3.6.0.dev1948%2Bgd8ede1a710/matplotlib-3.6.0.dev1948%2Bgd8ede1a710-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 18.5 MB/s eta 0:00:00
Requirement already satisfied: pyparsing<3.0.0,>=2.2.1 in /venv/lib/python3.10/site-packages (from matplotlib) (2.4.7)
Requirement already satisfied: cycler>=0.10 in /venv/lib/python3.10/site-packages (from matplotlib) (0.11.0)
Requirement already satisfied: packaging>=20.0 in /venv/lib/python3.10/site-packages (from matplotlib) (21.3)
Requirement already satisfied: fonttools>=4.22.0 in /venv/lib/python3.10/site-packages (from matplotlib) (4.31.2)
Requirement already satisfied: python-dateutil>=2.7 in /venv/lib/python3.10/site-packages (from matplotlib) (2.8.2)
Requirement already satisfied: kiwisolver>=1.0.1 in /venv/lib/python3.10/site-packages (from matplotlib) (1.4.2)
Requirement already satisfied: numpy>=1.19 in /venv/lib/python3.10/site-packages (from matplotlib) (1.22.3)
Requirement already satisfied: pillow>=6.2.0 in /venv/lib/python3.10/site-packages (from matplotlib) (9.0.1)
Requirement already satisfied: six>=1.5 in /venv/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)
Installing collected packages: matplotlib
Successfully installed matplotlib-3.6.0.dev1948+gd8ede1a710
(venv) root@ad729996566a:/# python -m pip show matplotlib
Name: matplotlib
Version: 3.6.0.dev1948+gd8ede1a710
Summary: Python plotting package
Home-page: https://matplotlib.org
Author: John D. Hunter, Michael Droettboom
Author-email: matplotlib-users@python.org
License: PSF
Location: /venv/lib/python3.10/site-packages
Requires: cycler, fonttools, kiwisolver, numpy, packaging, pillow, pyparsing, python-dateutil
Required-by: 

```

**Comment by ianhi** [comment]
ok I think the best possible option is:

```
pip install --upgrade --pre --index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple --extra-index-url https://pypi.org/simple matplotlib
```

this way you get the most up to date mpl dependencies (but download the deps from pypi)

**Comment by matthewfeickert** [comment]
> ok I think the best possible option is:
> 
> ```
> pip install --upgrade --pre --index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple --extra-index-url https://pypi.org/simple matplotlib
> ```
> 
> this way you get the most up to date mpl dependencies (but download the deps from pypi)

Yeah that works! I'll open a fix PR for the docs after dinner.

**Comment by tacaswell** [comment]
Whoops, it looks like we fixed the compatibility in with pyparsing 3 in https://github.com/matplotlib/matplotlib/pull/21501/files and removed the pinning in the CI requirement files but forgot to remove it in setup.py 🤦🏻

**Comment by tacaswell** [comment]
https://github.com/matplotlib/matplotlib/pull/22751 to fix the pyparsing issue.

**Comment by matthewfeickert** [comment]
PR #22752 to make the docs more robust.

**Comment by QuLogic** [comment]
[Duplicate uploads](https://github.com/matplotlib/matplotlib/runs/5781197219?check_suite_focus=true) also fail.

**Comment by tacaswell** [comment]
I think that is OK though?  

We have at least one build on main per day most days and I am not sure it is worth the logic to not get rejected on the days we do not (or on days when someone has pushed the button to run it early).

**Comment by matthewfeickert** [comment]
> [Duplicate uploads](https://github.com/matplotlib/matplotlib/runs/5781197219?check_suite_focus=true) also fail.

> I think that is OK though?
> 
> We have at least one build on main per day most days and I am not sure it is worth the logic to not get rejected on the days we do not (or on days when someone has pushed the button to run it early).

If no one minds I don't mind throwing in some additional logic to check just before the upload stage. This is pretty easy with `pip index versions` and some `sed`

```
$ python -m pip index \
    --index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple \
    --pre \
    versions matplotlib | \
  grep matplotlib | \
  sed 's/.*(\(.*\))/\1/'
WARNING: pip index is currently an experimental command. It may be removed/changed in a future release without prior warning.
3.6.0.dev1948+gd8ede1a710
```

So all you'd need to do would be something like

```console
$ LAST_NIGHTLY_VERSION="$(python -m pip index \
    --index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple \
    --pre \
    versions matplotlib | \
  grep matplotlib | \
  sed 's/.*(\(.*\))/\1/')"
WARNING: pip index is currently an experimental command. It may be removed/changed in a future release without prior warning.
$ echo "${LAST_NIGHTLY_VERSION}"
3.6.0.dev1948+gd8ede1a710
```

and then just check if that shows up as version of the wheels that just got downloaded from the GitHub Actions workflow artifact

```console
$ [ "$(find dist -type f -iname "matplotlib-${LAST_NIGHTLY_VERSION}*.whl" | wc --lines)" -gt "0" ]
$ echo $?
0
```

(maybe a more elegant way to do that but that's not the worst)

I'll make an Issue from this and I'm happy to PR it, unless you would prefer to avoid touching it as much as possible.

**Pull request merged**
